neps/nep-0364.md

```diff
@@ -0,0 +1,199 @@
+into multiple transactions.
+
+### Why is this design the best in the space of possible designs?
+
+Adding existing proved and vetted crypto crates into the runtime is a safe workaround. It will boost performance
+between 20-25x according to our benchmarks. This will both reduce operating costs significantly and will also
+enable the contract to verify all the signatures in one transaction, which will simplify the contract design.
+
+### What other designs have been considered and what is the rationale for not choosing them?
+
+One possible alternative would be to improve the runtime implementation so that it can compile WASM code to a sufficiently
+fast machine code. Even when it may not be as fast as LLVM native produced code it could still be acceptable for
+these types of use cases (CPU intensive functions) and will certainly avoid the need of adding host functions.
+The effort of adding such a feature will be significantly higher than adding these host functions one by one.
+But on the other side, it will decrease the need of including more host functions in the future.
+
+Another alternative is to deal with the high cost of computing/verifying these signatures in some other manner.
+Decreasing the overall cost of gas and increasing the limits of gas available to attach to the contract could be a possibility.
+Introducing such modification for some contracts, and not for some others can be rather arbitrary
+and not straightforward in the implementation, but an alternative nevertheless.
+
+## Specification
+
+This NEP aims to introduce the following host function:
+
+```rust
+/// Ed25519 is a public-key signature system with several attractive features
+///
+/// Proof of Stake Validator sets can contain different signature schemes.
+/// Ed25519 is one of the most used ones across blockchains, and hence it's importance to be added.
+/// For further reference, visit: https://ed25519.cr.yp.to
+///
+/// Verify an ED25519 signature given a message and a public key.
+/// # Returns
+/// - 1 meaning the boolean expression true to encode that the signature was properly verified
+/// - 0 meaning the boolean expression false to encode that the signature failed to be verified
+///
+/// # Cost
+///
+/// Each input can either be in memory or in a register. Set the length of the input to `u64::MAX`
+/// to declare that the input is a register number and not a pointer.
+/// Each input has a gas cost input_cost(num_bytes) that depends on whether it is from memory
+/// or from a register. It is either read_memory_base + num_bytes * read_memory_byte in the
+/// former case or read_register_base + num_bytes * read_register_byte in the latter. This function
+/// is labeled as `input_cost` below.
+///
+/// `input_cost(num_bytes_signature + num_bytes_message, num_bytes_public_key) +
+///  ed25519_verify_base + ed25519_verify_byte * (num_bytes_signature + num_bytes_message)`
```

Comment:
Double check that the cost matches what we actually charge. In partiuclar, I don't think we count signature lenght as a part of this component?

Reply:
[in the implementation](https://github.com/near/nearcore/pull/7165/files#diff-8df9ae8ecbf1224f7f000dd60280838a7399a49e00af2754cc41f5a4b1ac2640R1162) there's only variable cost for the message length. I can adjust it to match that.

Reply:
I've updated it